docs/swift/DesignDoc.md

```diff
@@ -0,0 +1,674 @@
+# Protocol Buffers in Swift
+
+## Objective
+
+This document describes the user-facing API and internal implementation of
+proto2 and proto3 messages in Apple’s Swift programming language.
+
+One of the key goals of protobufs is to provide idiomatic APIs for each
+language. In that vein, **interoperability with Objective-C is a non-goal of
+this proposal.** Protobuf users who need to pass messages between Objective-C
+and Swift code in the same application should use the existing Objective-C proto
+library. The goal of the effort described here is to provide an API for protobuf
+messages that uses features specific to Swift—optional types, algebraic
+enumerated types, value types, and so forth—in a natural way that will delight,
+rather than surprise, users of the language.
```

Comment:
One overall concern about this document:  It really needs to go into more detail on the user-facing API on the generated code.

For example, I would expect to see a section detailing such core methods as:
```
   public init(protobuf: [UInt8]) throws
   public init(json: String) throws
   public func serializeProtobuf() throws -> [UInt8]
   public func serializeJSON() throws -> String
   var isEmpty: Bool { get }
   var hashValue: Int { get }
```